### PR TITLE
feat(flow): render map&related tests

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Starting add the ability to "render" and execute dataflow Plan, this PR include simple rendering functions like `render_constant` and `render_mfp`(which renders a MapFilterProject) and comprehensive unit tests are added to test them.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Comment by killme2008** [comment]
@discord9 There are some typos, please fix them before inviting people to review.

File: src/flow/src/compute/render.rs [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+        // default to have a arrange with only future updates, so it can be empty if no temporal filter is applied
+        // as stream only sends current updates and etc.
+        let arrange = {
+            let arrangement = Arrangement::new();
+            ArrangeHandler::from(arrangement)
+        };
+
+        // This closure capture following variables:
+        let mfp_plan = MfpPlan::create_from(mfp).context(EvalSnafu)?;
+        let arrange_inner = arrange.clone_future_only().context(InvalidQuerySnafu {
```
Comment (waynexia) [inline src/flow/src/compute/render.rs:160]: prefer `with_context()` when constructing a non-static (like string allocation here) error

File: src/flow/src/compute/types.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+}
+
+/// A thread local error collector, used to collect errors during the evaluation of the plan
+///
+/// usually only the first error matters, but store all of them just in case
+#[derive(Default, Clone)]
+pub struct ErrCollector {
+    pub inner: Rc<RefCell<VecDeque<EvalError>>>,
```
Comment (waynexia) [inline src/flow/src/compute/types.rs:128]: Why `VecDeque` here. How will those stored errors be used later

**Review by waynexia — COMMENTED** [review]
Some style comments:

File: src/flow/src/compute/render.rs [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+            bundle.collection.stream.drop(self.df);
+            drop(bundle.arranged);
+        }
+        // The automatically generated "drop glue" which recursively calls the destructors of all the fields (including the now empty `input_collection`)
+    }
+}
+
+// There is a false positive in using `Vec<ScalarExpr>` as key
+#[allow(clippy::mutable_key_type)]
+impl<'referred, 'df> Context<'referred, 'df> {
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:66]: how about adding methods to control collection bundles, e.g. `insert_global(global id, collection)`, `insert_local(scope, local id, collection)`

File: src/flow/src/compute/render.rs [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+            self.local_scope.last_mut().unwrap()
+        };
+        local_scope.insert(id, value);
+        let ret = self.render_plan(*body)?;
+        Ok(ret)
+    }
+
+    /// render MapFilterProject, will only emit the `rows` once. Assume all incoming row's sys time being `now`` and ignore the row's stated sys time
+    /// TODO(discord9): schedule mfp operator to run when temporal filter need
+    pub fn render_mfp(
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:140]: The responsibilities and method names of these few methods don't seem to correspond well, making it difficult to grasp their side effects from their names.

Comment (zhongzc) [reply src/flow/src/compute/render.rs:140]: 这Can this function's implementation be split into smaller, more readable functions for non-experts?

File: src/flow/Cargo.toml [hunk 5]
```diff
@@ -16,7 +16,7 @@ common-telemetry.workspace = true
 common-time.workspace = true
 datatypes.workspace = true
 enum_dispatch = "0.3"
-hydroflow = "0.5.0"
+hydroflow = { git = "https://github.com/GreptimeTeam/hydroflow.git", rev = "ba2df44efd42b7c4d37ebefbf82e77c6f1d4cb94" }
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/flow/Cargo.toml:21]: > What changes are included in the forked version

Just a simple fork without extra modification, for getting our depend in our org.

File: src/flow/src/compute/types.rs (hunk 2 above)
Comment (discord9) [reply src/flow/src/compute/types.rs]: > Why `VecDeque` here. How will those stored errors be used later

The original idea was to use error collector to continous collect those non-lethal `EvalError` and output them to somewhere(i.e. log) while perserve their order

File: src/flow/src/compute/render.rs (hunk 4 above)
Comment (discord9) [reply src/flow/src/compute/render.rs]: > Can this function's implementation be split into smaller, more readable functions for non-experts?

split `render_mfp` into two function, and added more comment to explain

File: src/flow/src/compute/render.rs [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+        body: Box<Plan>,
+    ) -> Result<CollectionBundle, Error> {
+        let value = self.render_plan(*value)?;
+        let local_scope = if let Some(last) = self.local_scope.last_mut() {
+            last
+        } else {
+            self.local_scope.push(Default::default());
+            self.local_scope.last_mut().unwrap()
+        };
+        local_scope.insert(id, value);
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:160]: use `self.insert_local`

File: src/flow/src/compute/render.rs [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+        let arrange_inner = arrange
+            .clone_future_only()
+            .with_context(|| InvalidQuerySnafu {
+                reason: "Failed to clone future only arrangement".to_string(),
+            })?;
+        let as_of = self.compute_state.as_of.clone();
+        let err_collector = self.err_collector.clone();
+        // TODO(discord9): better way to schedule future run
+        let scheduler = self.compute_state.get_scheduler();
+        let cur_subgraph = scheduler.cur_subgraph.clone();
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:199]: `cur_subgraph` is a `None`, is it expected?

File: src/flow/src/compute/render.rs [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+    /// Eval `Let` operator, useful for assigning a value to a local variable
+    pub fn eval_let(
+        &mut self,
+        id: LocalId,
+        value: Box<Plan>,
+        body: Box<Plan>,
+    ) -> Result<CollectionBundle, Error> {
+        let value = self.render_plan(*value)?;
+
+        self.local_scope.push(Default::default());
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:159]: Is it expected to add a layer?

**Comment by discord9** [comment]
Yes, it's expected behavior, maybe I should change the layer from BTreemap to Tuple instead



---- Replied Message ----
| From | ***@***.***> |
| Date | 04/02/2024 18:33 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [GreptimeTeam/greptimedb] feat(flow): render map&related tests (PR #3581) |

@zhongzc commented on this pull request.

In src/flow/src/compute/render.rs:

> +
+        self.local_scope.push(Default::default());


Is it expected to add a layer?

File: src/flow/src/compute/render.rs [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+    for (mut row, _sys_time, diff) in input.into_iter() {
+        // this updates is expected to be only zero to two rows
+        let updates = mfp_plan.evaluate::<EvalError>(&mut row.inner, now, diff);
+        // TODO(discord9): refactor error handling
+        // Expect error in a single row to not interrupt the whole evaluation
+        let updates = updates
+            .filter_map(|r| match r {
+                Ok(r) => Some(((r.0, Row::empty()), r.1, r.2)),
+                Err(e) => {
+                    err_collector.push_err(e.0);
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:295]:
```suggestion
                Ok((key, ts, diff)) => Some(((key, Row::empty()), ts, diff)),
                Err((err, _ts, _diff)) => {
                    err_collector.push_err(err);
```

File: src/flow/src/compute/render.rs [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+        .into_iter()
+        .map(|((k, _v), t, d)| (k, t, d))
+        .collect_vec();
+    send.give(output);
+
+    let run_compaction = || {
+        arrange.write().set_compaction(now)?;
+        Ok(())
+    };
+    err_collector.run(run_compaction);
```
Comment (zhongzc) [inline src/flow/src/compute/render.rs:269]:
```suggestion
    // Deal with output:
    // 1. Read all updates that were emitted between the last time this arrangement had updates and the current time.
    // 2. Output the updates.
    // 3. Truncate all updates within that range.

    let from = arrange.read().last_compaction_time().map(|n| n + 1);
    let from = from.unwrap_or(repr::Timestamp::MIN);
    let output_kv = arrange.read().get_updates_in_range(from..=now);

    // the output is expected to be key -> empty val
    let output = output_kv
        .into_iter()
        .map(|((key, _v), ts, diff)| (key, ts, diff))
        .collect_vec();
    send.give(output);

    let run_compaction = || {
        arrange.write().compaction_to(now)?;
        Ok(())
    };
    err_collector.run(run_compaction);
```

**Review by zhongzc — APPROVED** [review]

File: src/flow/src/plan.rs [hunk 11]
```diff
@@ -44,12 +44,19 @@ pub enum Plan {
     Get { id: Id },
     /// Create a temporary collection from given `value``, and make this bind only available
```
Comment (waynexia) [inline src/flow/src/plan.rs:45]:
```suggestion
    /// Create a temporary collection from given `value`, and make this bind only available
```

File: src/flow/src/compute/render.rs [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+                *cnt_inner.borrow_mut() += data.iter().map(|v| v.len()).sum::<usize>();
+            },
+        );
+        ctx.df.run_available();
+        assert_eq!(*cnt.borrow(), 3);
+        ctx.df.run_available();
+        assert_eq!(*cnt.borrow(), 3);
+    }
+
+    /// a simple example to show how to use source and sink
```
Comment (waynexia) [inline src/flow/src/compute/render.rs:511]: 👍

**Review by waynexia — APPROVED** [review]
I went through the code and test and haven't found any obvious problem 🎸 `example_source_sink` is a great example, we can consider linking it in the code document somewhere

**Pull request merged**